modules/dnn/src/cuda4dnn/csl/tensor_ops.hpp

```diff
@@ -0,0 +1,384 @@
+                convDesc, algo, scratchpad,
+                filterDesc, filters.get(),
+                inputTensorDesc, input.get(),
+                1.0, 0.0, outputTensorDesc, output.get()
+            );
+        }
+
+    private:
+        cudnn::Handle cudnnHandle;
+        TensorDescriptor inputTensorDesc, outputTensorDesc;
+        FilterDescriptor filterDesc;
+        ConvolutionDescriptor convDesc;
+        ConvolutionAlgorithm algo;
+    };
+
+    template <class T>
+    class Pooling {
+        using TensorDescriptor = cudnn::TensorDescriptor<T>;
+        using PoolingDescriptor = cudnn::PoolingDescriptor;
+
+    public:
+        enum class rounding_type {
+            FLOOR,
+            CEILING
+        };
+
+        using pooling_type = PoolingDescriptor::pooling_type;
+
+        struct params_type {
+            std::vector<std::size_t> input_shape;
+
+            std::vector<std::size_t> window_size;
+            std::vector<std::size_t> padding;
+            std::vector<std::size_t> stride;
+
+            rounding_type rounding_mode;
+            pooling_type type;
+        };
+
+        Pooling() = default;
+        Pooling(const Pooling&) = delete;
+        Pooling(Pooling&&) = default;
+        Pooling(cudnn::Handle handle, const params_type& params) {
+            cudnnHandle = std::move(handle);
+
+            inputTensorDesc = TensorDescriptor(params.input_shape);
+            poolingDesc = PoolingDescriptor(params.window_size, params.padding, params.stride, params.type);
+
+            const auto& input_shape = params.input_shape;
+            std::vector<std::size_t> output_shape;
+            output_shape.assign(std::begin(input_shape), std::end(input_shape));
+
+            const auto& window_size = params.window_size;
+            const auto& padding = params.padding;
+            const auto& stride = params.stride;
+
+            bool ceil = params.rounding_mode == rounding_type::CEILING;
+            for (int i = 0; i < window_size.size(); i++) {
+                double axis_sz = (input_shape[i + 2] + 2 * padding[i] - window_size[i]) / double(stride[i]) + 1;
+                output_shape[i + 2] = ceil ? std::ceil(axis_sz) : std::floor(axis_sz);
+
+                /* check if the last pooling window starts in the valid region */
+                if (padding[i]) {
+                    if ((output_shape[i + 2] - 1) * stride[i] >= input_shape[i + 2] + padding[i])
+                        output_shape[i + 2]--;
+                }
+            }
+
+            if (!ceil)
+            {
+                /* we must agree with cuDNN if we used floor */
+                std::vector<int> output_dim;
+                getPoolingForwardOutputDim(poolingDesc, inputTensorDesc, output_dim);
+                CV_Assert(std::equal(std::begin(output_dim), std::end(output_dim), std::begin(output_shape)));
+                CV_UNUSED(output_dim);
+            }
+
+            outputTensorDesc = TensorDescriptor(output_shape);
```

Comment:
There is undefined behaviour here. cuDNN rounds down when it encounters fractional output dimensions. To allow rounding upwards, I am creating a tensor descriptor of required outputs instead of using the output descriptor provided by cuDNN. This surprisingly works but makes use of undocumented implementation details of cuDNN (or it might have been documented and I might be unaware of it).

This code stays for now. I'll fix this when I write pooling kernels for deconvolution networks.